IBC.md

```diff
@@ -305,10 +308,101 @@ pub struct IbcPacket {
+messages) to be reverted when the packet is rejected, but the transaction to
+properly commit an acknowledgement with encoded error. In other words, this "IBC
+Handler" will error and revert, but the "IBC Router" must succeed and commit an
+acknowledgement message (that can be parsed by the sending chain as an error).
+
+The atomicity issue was first
+[analyzed in the Cosmos SDK implementation](https://github.com/cosmos/ibc-go/issues/68)
+and refined into
+[changing semantics of the OnRecvPacket SDK method](https://github.com/cosmos/ibc-go/issues/91),
+which was
+[implemented in April 2021](https://github.com/cosmos/ibc-go/pull/107), likely
+to be released with Cosmos SDK 0.43 or 0.44. Since we want the best,
+future-proof interface for contracts, we will use an approach inspired by that
+work, and add an adapter in `wasmd` until we can upgrade to a Cosmos SDK version
+that implements this.
+
+After quite some
+[discussion on how to encode the errors](https://github.com/CosmWasm/cosmwasm/issues/762),
+we identified a number of issues implementing this idea with the CosmWasm model:
+
+1. We only return `Ok(Response)` or `Err(String)`. There is no way to return an
+   arbitrary binary encoded message in the error variant.
+2. If we return `messages` to be dispatched, they may fail with an error, which
+   is returned directly to the x/wasm handler. Obviously an error from the bank
+   handler will not be in the proper IBC Acknowledgement format.
+3. We also realized that in some cases, the original `ibc_packet_receive` will
+   not even have the information needed to format the success response, so we
+   will need some delayed response-builder in this case also.
+
+eg. In the ICS27 spec, they define a packet to register a new account, the
+acknowledgement of that packet should contain the address of the new account on
+the receiving chain. If we implemented this in CosmWasm, the contract would have
+to return a submessage of `WasmMsg::Instantiate{}` and only get the proper
+address for the response in `reply`. Thus, we need to have some way to fully
+build the success acknowledgement after the fact as well.
+
+With this in mind, we designed the following workflow for the `x/wasm` runtime:
+
+1. Wrap the storage with a "local cache", so we don't immediately write out
+   changes
+2. Call `ibc_receive_packet`, get either `Err(String)` or
+   `Ok(Response + acknowledgement)`. If `Err`, jump to 4.
+3. Execute any `submessages` and `messages` returned with this response. If any
+   return `Err`, jump to 4.
+4. `receiveResult` is `Err(String)` if any of the above errored, or `Ok(Binary)`
+   with the original acknowledgement bytes if we have the success case.
+5. Call `ibc_encode_acknowledgement` with the `receiveResult`. The return value
+   is the acknowledgement packet to send to the calling chain. This should never
+   error.
+6. If `ReceiveResult` is `Ok`, commit this "storage cache". If it is `Err`,
+   revert those changes.
```

Comment:
🤔 Do I understand this correct that this step let the contract handle the error case on messages and control the TX scope? The process I was assuming would revert the TX on error in 3) but can persist new state in 5).

Reply:
The future IBC design (linked above, here again: https://github.com/cosmos/ibc-go/issues/91) explicitly makes the entire app execution (x/wasm) atomic - to be rolled back when app error. But not touching the IBC part of the transaction and successfully committing a "error acknowledgement" in the IBC Keeper.

I feel it is pointless to design something other than what they are doing, unless you want to re-open that issue in the ibc repo. As in the mid-term, we will just use those semantics in `x/wasm` and all other modules will use them as well.